ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -306,22 +314,71 @@ private void recover(MessageStoreRecovery recovery) throws StoreException, IOExc
    * @throws IOException if an end offset could not be set due to I/O error.
    */
   private void setEndOffsets() throws IOException {
-    for (Offset segmentStartOffset : indexes.keySet()) {
-      Offset nextIndexSegmentStartOffset = indexes.higherKey(segmentStartOffset);
+    for (Offset segmentStartOffset : indexesRef.get().keySet()) {
+      Offset nextIndexSegmentStartOffset = indexesRef.get().higherKey(segmentStartOffset);
       if (nextIndexSegmentStartOffset == null || !segmentStartOffset.getName()
           .equals(nextIndexSegmentStartOffset.getName())) {
         // this is the last index segment for the log segment it refers to.
-        IndexSegment indexSegment = indexes.get(segmentStartOffset);
+        IndexSegment indexSegment = indexesRef.get().get(segmentStartOffset);
         LogSegment logSegment = log.getSegment(segmentStartOffset.getName());
         logSegment.setEndOffset(indexSegment.getEndOffset().getOffset());
       }
     }
   }
 
   /**
+   * @return the map of {@link Offset} to {@link IndexSegment} instances.
+   */
+  ConcurrentSkipListMap<Offset, IndexSegment> getIndexSegments() {
+    return indexesRef.get();
+  }
+
+  /**
+   * Atomically adds {@code segmentFilesToAdd} to and removes {@code segmentsToRemove} from the map of {@link Offset} to
+   * {@link IndexSegment} instances.
+   * @param segmentFilesToAdd the backing files of the {@link IndexSegment} instances to add.
+   * @param segmentsToRemove the start {@link Offset} of {@link IndexSegment} instances to remove.
+   * @throws IllegalArgumentException if any {@link IndexSegment} that needs to be added or removed has an offset higher
+   * than that in the journal.
+   * @throws StoreException if an {@link IndexSegment} instance cannot be created for the provided files to add.
+   */
+  void changeIndexSegments(List<File> segmentFilesToAdd, Set<Offset> segmentsToRemove) throws StoreException {
+    TreeMap<Offset, IndexSegment> segmentsToAdd = new TreeMap<>();
+    Offset journalFirstOffset = journal.getFirstOffset();
+    for (File indexSegmentFile : segmentFilesToAdd) {
+      IndexSegment indexSegment = new IndexSegment(indexSegmentFile, true, factory, config, metrics, journal);
+      if (indexSegment.getEndOffset().compareTo(journalFirstOffset) > 0) {
+        throw new IllegalArgumentException("One of the index segments has an end offset " + indexSegment.getEndOffset()
+            + " that is higher than the first offset in the journal " + journalFirstOffset);
+      }
+      segmentsToAdd.put(indexSegment.getStartOffset(), indexSegment);
+    }
+
+    indexesChangeLock.lock();
+    try {
+      ConcurrentSkipListMap<Offset, IndexSegment> oldIndexSegments = indexesRef.get();
+      ConcurrentSkipListMap<Offset, IndexSegment> newIndexSegments = new ConcurrentSkipListMap<>();
+      for (Map.Entry<Offset, IndexSegment> entry : oldIndexSegments.entrySet()) {
+        if (!segmentsToRemove.contains(entry.getKey())) {
```

Comment:
I don't see why a lock is needed. Doesn't the atomic reference suffice? Couldn't this method simply do the following:
- do the checks for whether any of the entries (in either `segmentsToAdd` or `segmentsToRemove`) is in the  trailing (journal-covered) part of the Index, if so throw. If not, it cannot in the future either since the trailing part only moves forward.
- Make the switch atomically as follows (there are some optimizations below that you can use even if you do need the lock):
```java
newIndexSegments = new ConcurrentSkipListMap<>(oldIndexSegments); // construct from the old.
newIndexSegments.putAll(segmentsToAdd);
newIndexSegments.keySet().removeAll(segmentsToRemove);
indexesRef.set(newIndexSegments);
```
I don't see a need for an `indexesChangeLock` in this flow and `addToIndex()` does not need it either (see my comment there). I get it if you think the changes that really need this is coming later on in compaction. Would like to hear anyway as to why the atomic reference change would not suffice.